File: leetcode/3-longest-substring-without-repeating-characters/main.go
package main

import "fmt"

/*
	1st approach:
	- 2 pointers + hashtable
	- when we see a precedented char, we update the result and remove the items which values(indeces) < slow pointer

	Time	O(n^2)
	Space	O(n)
	beats 22.82%
*/
func lengthOfLongestSubstring(s string) int {
	m := make(map[string]int)
	last := 0
	res := 0
	for i := 0; i < len(s); i++ {
		c := string(s[i])
		if _, x := m[c]; !x {
			m[c] = i
		} else {
			res = max(res, i-last)
			recur := m[c]
			last = recur + 1
			m[c] = i
			for k, _ := range m {
				if m[k] < last {
					delete(m, k)
				}
			}
		}
	}
	res = max(res, len(s)-last)
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

/*
	2nd approach:
	- optimize the 1st approach, hashtable
	- but we dont need to remove the item from hashtable
	- instead we can just ignore the item which index < slow pointer

	Time	O(n)
	Space	O(n)
	beats 69.67%
*/
func lengthOfLongestSubstring1(s string) int {
	m := make(map[string]int)
	last := 0
	res := 0
	for i := 0; i < len(s); i++ {
		c := string(s[i])
		if _, x := m[c]; !x {
			m[c] = i
		} else {
			recur := m[c]
			if recur >= last {
				res = max(res, i-last)
				last = recur + 1
			}
			m[c] = i
		}
	}
	res = max(res, len(s)-last)
	return res
}

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb"))
	fmt.Println(lengthOfLongestSubstring("bbbbb"))
	fmt.Println(lengthOfLongestSubstring("pwwkew"))
	fmt.Println(lengthOfLongestSubstring("abcdef"))
	fmt.Println(lengthOfLongestSubstring("abcbdef"))
	fmt.Println(lengthOfLongestSubstring("abcbdee"))
	fmt.Println(lengthOfLongestSubstring("abb"))
	fmt.Println(lengthOfLongestSubstring("ab"))
	fmt.Println(lengthOfLongestSubstring("a"))

	fmt.Println("---")

	fmt.Println(lengthOfLongestSubstring1("abcabcbb"))
	fmt.Println(lengthOfLongestSubstring1("bbbbb"))
	fmt.Println(lengthOfLongestSubstring1("pwwkew"))
	fmt.Println(lengthOfLongestSubstring1("abcdef"))
	fmt.Println(lengthOfLongestSubstring1("abcbdef"))
	fmt.Println(lengthOfLongestSubstring1("abcbdee"))
	fmt.Println(lengthOfLongestSubstring1("abb"))
	fmt.Println(lengthOfLongestSubstring1("ab"))
	fmt.Println(lengthOfLongestSubstring1("a"))
}

File: leetcode/3-longest-substring-without-repeating-characters/main.py
from collections import Counter
"""
    2nd approach: sliding window + hashtable
    - similar to lc3, 159, 340, 904
	- in each iteration, if we meet a 2-occurence character
        - we move the slow pointer to the left
        - and increment the count of the character on the index where slow pointer points to
        - continue to move the slow pointer until the current character's count becomes 1

	Time	O(2n)
	Space	O(n)
    60 ms, faster than 52.92%
"""


class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        res = 0
        counter = Counter()
        j = 0
        for i in range(len(s)):
            c = s[i]
            counter[c] += 1
            while counter[c] > 1:
                left = s[j]
                j += 1
                counter[left] -= 1
            res = max(res, i - j + 1)
        return res


a = "abcabcbb"
print(Solution().lengthOfLongestSubstring(a))

a = "bbbbb"
print(Solution().lengthOfLongestSubstring(a))

a = "pwwkew"
print(Solution().lengthOfLongestSubstring(a))

a = "abcdef"
print(Solution().lengthOfLongestSubstring(a))

a = "abcbdef"
print(Solution().lengthOfLongestSubstring(a))

a = "abcbdee"
print(Solution().lengthOfLongestSubstring(a))

a = "abb"
print(Solution().lengthOfLongestSubstring(a))

a = "ab"
print(Solution().lengthOfLongestSubstring(a))

a = "a"
print(Solution().lengthOfLongestSubstring(a))

print("-----")

"""
    followup: print that substring

    Time	O(2n)
	Space	O(n)
    64 ms, faster than 46.35%
"""


class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        ht = Counter()
        j = 0
        res_left, res_right = 0, 0
        for i in range(len(s)):
            c = s[i]
            ht[c] += 1
            while ht[c] > 1:
                left = s[j]
                ht[left] -= 1
                j += 1
            if i - j + 1 > res_right - res_left + 1:
                res_left = j
                res_right = i
        return s[res_left: res_right+1]


a = "abcabcbb"
print(Solution().lengthOfLongestSubstring(a))

a = "bbbbb"
print(Solution().lengthOfLongestSubstring(a))

a = "pwwkew"
print(Solution().lengthOfLongestSubstring(a))

a = "abcdef"
print(Solution().lengthOfLongestSubstring(a))

a = "abcbdef"
print(Solution().lengthOfLongestSubstring(a))

a = "abcbdee"
print(Solution().lengthOfLongestSubstring(a))

a = "abb"
print(Solution().lengthOfLongestSubstring(a))

a = "ab"
print(Solution().lengthOfLongestSubstring(a))

a = "a"
print(Solution().lengthOfLongestSubstring(a))
